### PR TITLE
Add ShareesAPI E-mail search

### PR DESCRIPTION
**Pull request by rullzer**
- Allow to search for SHARE_TYPE_EMAIL (4)
- Added tests

Needed for https://github.com/nextcloud/server/pull/1801

@nickvergessen as discussed
@schiessle this is probabaly also usefull for https://github.com/nextcloud/server/pull/657

Please review

CC: @LukasReschke @MorrisJobke @icewind1991 @blizzz

### CODE REVIEW

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 1]
```diff
@@ -403,6 +405,72 @@ protected function fixRemoteURL($remote) {
...
+			$emails = $contact['EMAIL'];
+			if (!is_array($emails)) {
+				$emails = [$emails];
+			}
+
+			foreach ($emails as $email) {
+				if (strtolower($search) === $contact['FN'] ||
+					strtolower($search) === strtolower($email)
+				) {
+					if (strtolower($search) === $email) {
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: missing strtolower ?

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 2]
```diff
@@ -403,6 +405,72 @@ protected function fixRemoteURL($remote) {
 	}
 
 	/**
+	 * @param string $search
+	 */
+	protected function getEmails($search) {
+		$this->result['emails'] = [];
+		$this->result['exact']['users'] = [];
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: emails not users

File: apps/files_sharing/lib/Controller/ShareesAPIController.php [hunk 3]
```diff
@@ -403,6 +405,72 @@ protected function fixRemoteURL($remote) {
...
+					}
+
+					$this->result['exact']['emails'][] = [
+						'label' => $contact['FN'],
+						'value' => [
+							'shareType' => Share::SHARE_TYPE_EMAIL,
+							'shareWith' => $email,
+						],
+					];
+				} else {
```
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: `else if (!$this->shareeEnumeration)`
and kill the if after the loop?

File: apps/files_sharing/lib/Controller/ShareesAPIController.php (hunk 2 above)
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: emails not users

File: apps/files_sharing/lib/Controller/ShareesAPIController.php (hunk 1 above)
Comment (nickvergessen) [inline apps/files_sharing/lib/Controller/ShareesAPIController.php]: missing strtolower

**Review by nickvergessen — DISMISSED** [review]

**Comment by rullzer** [comment]
@nickvergessen addressed. All tests pass now.

**Comment by nickvergessen** [comment]
👍

**Comment by rullzer** [comment]
@MorrisJobke @LukasReschke @schiessle mind a review?

**Comment by MorrisJobke** [comment]
:+1:

**Comment by schiessle** [comment]
@rullzer I fear that this will conflict with https://github.com/nextcloud/server/pull/657 because here I only want to search for emails if the share-by-mail provider is loaded. At the point in the code where the search happens we can't distinguish if the request comes from the share auto-complete or from the email field for link shares. Wouldn't it be possible to use something else here than the "SHARE_TYPE_EMAIL"?

**Comment by rullzer** [comment]
Uhm wel only if we would introduce another share type

**Comment by schiessle** [comment]
do we really need a share provider or share type just to auto complete email addresses? I mean the email addresses added for link shares never become real shares.

**Comment by rullzer** [comment]
No but we need to signal the sharee api that we want autocomplete of a certain type. And that uses the share types

**Comment by schiessle** [comment]
The Share manager should handle shares. Asking for a "fake share type" just to auto-complete email addresses sounds wrong to me. I think this should be done either by a different API (ContactsMananger API) which is independent from sharing (because auto completing email addresses has nothing to do with sharing and could be done in many other places as well) or if it really needs to be part of sharing it should be part of the "link share provider/type" and not a separate share type.

I think you can judge best what can be done, because you know the sharing code best. But if we really need to use a "fake share type" we need two. So either you or I need to move to a new one.

**Comment by nickvergessen** [comment]
> I think this should be done either by a different API (ContactsMananger API) which is independent from sharing (because auto completing email addresses has nothing to do with sharing and could be done in many other places as well)

For the record I tend to agree. But lets continue discussion in https://github.com/nextcloud/server/issues/1916

**Comment by rullzer** [comment]
Yeah probabaly solved in #1916 is in.

**Pull request merged**